leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java

```diff
@@ -39,40 +39,54 @@
     private final ConcurrentMap<String, BlockingQueue<QueuedRequest>> requestQueueMap = new ConcurrentHashMap<>();
 
     @Override
-    public void add(QueuedRequest entity) {
+    public void add(final QueuedRequest entity) {
         LOG.debug("Add entity {}", entity);
-        String endpoint = entity.getEndpoint();
-        BlockingQueue<QueuedRequest> requestQueue = getMessageQueueForEndpoint(endpoint);
-        requestQueue.add(entity);
-        requestQueueMap.putIfAbsent(endpoint, requestQueue);
+        final String endpoint = entity.getEndpoint();
+        final BlockingQueue<QueuedRequest> requestQueue = new LinkedBlockingQueue<>();
+
+        // If two Threads want to add a entity at the exact instant, each will
+        // retrieve its own instance of
+        // Queue and add the entity to the respective instances. But only one of
+        // them wins the putIfAbsent method.
+        // To not lose messages, this check will prevent.
+        final BlockingQueue<QueuedRequest> allreadyExisting = requestQueueMap.putIfAbsent(endpoint, requestQueue);
+        if (allreadyExisting != null) {
+            allreadyExisting.add(entity);
+        } else {
+            requestQueue.add(entity);
+        }
     }
 
     @Override
-    public boolean isEmpty(String endpoint) {
-        return getMessageQueueForEndpoint(endpoint).isEmpty();
+    public boolean isEmpty(final String endpoint) {
+        if (queueExist(endpoint)) {
+            return requestQueueMap.get(endpoint).isEmpty();
+        }
+        return true;
+
```

Comment:
Ok, that is  possible, you are right.
My thought was not to have so much duplicate code, therefore the method. I will change this and remove the method.